crates/fullnode/src/runner.rs

```diff
@@ -186,467 +195,625 @@ where
-                        .get_l2_range_by_l1_height(SlotNumber(end_l1_height))?
-                    {
-                        Some((_, end_l2_height)) => end_l2_height,
-                        None => {
-                            tracing::warn!(
-                            "Sequencer commitment verification: L1 L2 connection does not exist. L1 height = {}. Skipping commitment.",
-                            end_l1_height
-                            );
-                            continue;
-                        }
-                    };
-
-                    let range_end = BatchNumber(end_l2_height.0 + 1);
-                    // Traverse each item's field of vector of transactions, put them in merkle tree
-                    // and compare the root with the one from the ledger
-                    let stored_soft_batches: Vec<StoredSoftBatch> = self
-                        .ledger_db
-                        .get_soft_batch_range(&(start_l2_height..range_end))?;
-
-                    let soft_batches_tree = MerkleTree::<Sha256>::from_leaves(
-                        stored_soft_batches
-                            .iter()
-                            .map(|x| x.hash)
-                            .collect::<Vec<_>>()
-                            .as_slice(),
-                    );
-
-                    if soft_batches_tree.root() != Some(sequencer_commitment.merkle_root) {
-                        tracing::warn!(
-                            "Merkle root mismatch - expected 0x{} but got 0x{}. Skipping commitment.",
-                            hex::encode(
-                                soft_batches_tree
-                                    .root()
-                                    .ok_or(anyhow!("Could not calculate soft batch tree root"))?
-                            ),
-                            hex::encode(sequencer_commitment.merkle_root)
-                        );
-                    } else {
-                        self.ledger_db.update_commitments_on_da_slot(
-                            soft_batch.da_slot_height,
-                            sequencer_commitment.clone(),
-                        )?;
-
-                        for i in start_l1_height..=end_l1_height {
-                            self.ledger_db.put_soft_confirmation_status(
-                                SlotNumber(i),
-                                SoftConfirmationStatus::Finalized,
-                            )?;
-                        }
+                },
+                Some((l2_height, l2_block)) = l2_rx.recv() => {
+                    let l1_block = get_da_block_at_height(&self.da_service, l2_block.da_slot_height).await?;
+                    if let Err(e) = self.process_l2_block(l2_height, l2_block, l1_block).await {
+                        error!("Could not process L2 block: {}", e);
                     }
-                }
-
-                cur_l1_block = Some(filtered_block);
+                },
```

Comment:
two things here:

1. assume there is 100 L2 blocks on an L1 block, we fetch the L1 block 100 times, this used to be just once. (on l1 height changed)
2. L1 syncer is already fetching the L1 blocks, can't we have a shared L1 block fetcher that fetches a given height/hash just once and serves both tasks?

Reply:
Handled

Reply:
Added LRU cache